ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/impl/InMemoryLedgerReaderWriter.scala

```diff
@@ -0,0 +1,139 @@
+private[impl] case class LogEntry(entryId: DamlLogEntryId, payload: Array[Byte])
+
+private[impl] case class InMemoryState(
+    log: mutable.Buffer[LogEntry] = ArrayBuffer[LogEntry](),
+    state: mutable.Map[ByteString, DamlStateValue] = mutable.Map.empty)
+
+class InMemoryLedgerReaderWriter(
+    ledgerId: LedgerId = Ref.LedgerString.assertFromString(UUID.randomUUID.toString),
+    val participantId: ParticipantId)(implicit executionContext: ExecutionContext)
+    extends LedgerWriter
+    with LedgerReader {
+
+  private val engine = Engine()
+
+  private val currentState = InMemoryState()
+
+  private val StartOffset: Int = 0
+
+  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] =
+    Future {
+      val submission = Envelope
+        .openSubmission(envelope)
+        .getOrElse(throw new IllegalArgumentException("Not a valid submission in envelope"))
+      currentState.synchronized {
+        val stateInputs: Map[DamlStateKey, Option[DamlStateValue]] =
+          submission.getInputDamlStateList.asScala
+            .map(key => key -> currentState.state.get(key.toByteString))(breakOut)
+        val entryId = allocateEntryId
+        val (logEntry, damlStateUpdates) =
+          KeyValueCommitting.processSubmission(
+            engine,
+            entryId,
+            currentRecordTime(),
+            LedgerReader.DefaultTimeModel,
+            submission,
+            participantId,
+            stateInputs
+          )
+        verifyStateUpdatesAgainstPreDeclaredOutputs(damlStateUpdates, entryId, submission)
+        val stateUpdates = damlStateUpdates.toSeq.map {
+          case (damlStateKey, value) => damlStateKey.toByteString -> value
+        }
+        currentState.log += LogEntry(entryId, Envelope.enclose(logEntry).toByteArray)
+        currentState.state ++= stateUpdates
+        dispatcher.signalNewHead(currentState.log.size)
+      }
+      SubmissionResult.Acknowledged
+    }
+
+  private def verifyStateUpdatesAgainstPreDeclaredOutputs(
+      actualStateUpdates: Map[DamlStateKey, DamlStateValue],
+      entryId: DamlLogEntryId,
+      submission: DamlSubmission): Unit = {
+    val expectedStateUpdates = KeyValueCommitting.submissionOutputs(entryId, submission)
+    if (!(actualStateUpdates.keySet subsetOf expectedStateUpdates)) {
+      val unaccountedKeys = actualStateUpdates.keySet diff expectedStateUpdates
+      sys.error(
+        s"CommitActor: State updates not a subset of expected updates! Keys [$unaccountedKeys] are unaccounted for!")
+    }
+  }
+
+  override def events(offset: Option[Offset]): Source[LedgerRecord, NotUsed] =
+    dispatcher
+      .startingAt(
+        offset
+          .map(_.components.head.toInt)
+          .getOrElse(StartOffset),
+        OneAfterAnother[Int, List[LedgerRecord]](
+          (index: Int, _) => index + 1,
+          (index: Int) => Future.successful(List(getLogEntry(index)))
+        )
+      )
+      .collect {
```

Comment:
Since no "filtering" is happening in this `collect`, you can call `mapConcat` directly.

Reply:
Changed.